CONTRIBUTING.md

```diff
@@ -0,0 +1,437 @@
+
+## API conventions
+This section covers Fabric API guidelines for common patterns.
+
+### Backwards compatibility
+Fabric API makes strong backwards compatibility guarantees, by which contributors must abide. **Modders should not need to update their source code when they update Fabric API**, with the following exceptions:
+
+- APIs might be broken due to **related** changes in Minecraft updates.
+    - Preserving backwards compatibility in these cases will be considered if possible.
+        - Deprecated code will usually be removed if it requires non-trivial updates.
+    - This is limited to code directly affected by the Minecraft version update. Minecraft making breaking changes in some areas of its API is not a reason to break unrelated parts of Fabric API.
+- Experimental modules might be broken if necessary.
+    - Changes that affect compilation but not the resulting bytecode are allowed in experimental modules, for example some generic changes.
+    - Even if an API is marked as experimental, avoid breaking the ecosystem of mods using it.
+        - Changes that will cause hard crashes in many mods when they are released should be delayed until the next majorly breaking Minecraft version update.
+- There might not be a way around a breaking change.
+    - If a major oversight is noticed in an API, it might be broken to address the oversight. Such cases require careful consideration, and should always be discussed with the team beforehand.
+    - Remember that deprecating code is always an option.
+
+### Targeted Minecraft versions
+- Fabric API does not have strict Minecraft version support policies, but rather supports what is feasible and what the community is interested in.
+    - In particular, the latest stable Minecraft version and the latest subsequent snapshot or pre-release version are always supported.
+    - Which features go in older stable versions is a tradeoff between how easy it is, how many mods are still using that version, and what the community is willing to contribute.
+- New features can be targeted at any supported version.
+    - Maintainers will take care of cherry-picking the feature to the other branches when applicable.
+    - In doubt, prefer PR'ing to newer stable versions.
+- Backporting PRs for older versions will generally be accepted, depending on how many changes were required.
+
+### Discouraged API design patterns
+
+- Avoid the old `Api.INSTANCE` pattern in favor of static methods.
+    - This pattern may be used if the backing implementation may be changed by another mod (such as the renderer api). Even then there should be static methods to access the standard api facilities.
+- Avoid unnecessary use of generics.
+    - Unless Vanilla Minecraft mandates the use of generics, or a good reason exists to use generics.
+    - In particular, use of self types should be avoided.
```

Comment:
Why?

Reply:
Well, they lead to many problems and should be avoided for that reason.

Reply:
> many problems

...that doesn't really answer anything.

Reply:
I don't have a specific example in mind, but I had issues when using them in the past in AE2. `IAEStack<T extends IAEStack<T>>` used to be generic hell.

Reply:
Self types are fine - for example vanilla's loot builders use them without them being generic hell. Using generics (including self types) is one thing, going overboard with them is another.

Reply:
I removed this line since it's already contained in `Avoid unnecessary use of generics.`